scalafix-core/shared/src/main/scala/scalafix/internal/rule/OrphanImplicits.scala

```diff
@@ -0,0 +1,66 @@
+package scalafix.internal.rule
+
+import scala.meta._
+import scalafix.internal.util.SymbolOps
+import scalafix.lint.LintCategory
+import scalafix.rule.RuleCtx
+import scalafix.util.SemanticdbIndex
+import scalafix.{LintMessage, Rule}
+
+final case class OrphanImplicits(index: SemanticdbIndex)
+    extends Rule("OrphanImplicits") {
+
+  override def description: String =
+    "Linter that reports an error on implicit instances of shape F[G] " +
+      "that do not belong in the companion objects of F or G."
+
+  private lazy val errorCategory: LintCategory =
+    LintCategory.error("Orphan implicits should be avoided")
+
+  override def check(ctx: RuleCtx): Seq[LintMessage] = {
+
+    def handleImplicit(
+        tpe: Type,
+        obj: Symbol.Global,
+        pos: Position): Option[LintMessage] = {
+
+      val tpeSymbols = tpe match {
+        case Type.Apply(t, Seq(arg)) =>
+          Seq(index.symbol(t), index.symbol(arg)).flatten
+        case _ => Seq.empty
+      }
+
+      if (tpeSymbols.nonEmpty && !tpeSymbols.exists(
+          SymbolOps.isSameNormalized(obj, _)
+        )) {
+        Some(
+          errorCategory
+            .at(
+              s"""Orphan implicits are not allowed.
+                 |You should put this definition to one of the following objects:
+                 |${tpeSymbols.mkString(", ")}
+                """.stripMargin,
+              pos
+            ))
+      } else {
+        None
+      }
+    }
+
+    ctx.tree.collect {
+      case Defn.Object(_, name, templ) =>
+        index.symbol(name) match {
+          case Some(obj @ Symbol.Global(_, _)) =>
+            templ.stats.collect {
+              case t @ Defn.Val(mods, _, Some(tpe), _)
```

Comment:
This won't catch implicits with inferred types

```scala
object Foo {
  implicit val x = List(1)
}
```

The robust way to implement this is to query the type of a definition is to use the `.resultType` extension method as is done here https://github.com/scalacenter/scalafix/blob/e46a0401c78d955a9ee218555cfffc7eff2c8230/scalafix-core/shared/src/main/scala/scalafix/internal/rule/ExplicitResultTypes.scala#L75-L78

`.resultTypes` will soon be refactored to use `Denotation.tpe: semanticdb3.Type`, but I can take care of that refactoring once that happens.